[ProxyBridge] Fix proxy classnames generation

xphere opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | N/A |
| License | MIT |
| Doc PR | N/A |

When having many lazy services based on the same class, proxy classnames collision is more prone to happen on cache regeneration.

The problem comes from [`ProxyDumper`](https://github.com/symfony/symfony/blob/2.3/src/Symfony/Bridge/ProxyManager/LazyProxy/PhpDumper/ProxyDumper.php#L112) usage of [`spl_object_hash`](http://php.net/manual/en/function.spl-object-hash.php), because object hashes may be reused for others when destroyed.

To solve this, I added more entropy with an optional salt to allow multiple passes of the `ProxyDumper`, each time with a value coming from the filepath of the cache we're building, so no collision could happen.

This applies from `symfony/2.3` to `symfony/2.7`.
Related tests are passing.

fabpot commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -683,7 +683,8 @@ protected function dumpContainer(ConfigCache $cache, ContainerBuilder $container
         if (class_exists('ProxyManager\Configuration')) {
-            $dumper->setProxyDumper(new ProxyDumper());
+            $salt = md5((string) $cache);
+            $dumper->setProxyDumper(new ProxyDumper($salt));
```
Can you inline the var?

xphere replied:
Rebased and fixed

xphere committed:
add more entropy to generated classnames

- use proxy destination filename to create and md5 salt
- apply fabbot patch
- inline salt generation

sstok commented on src/Symfony/Component/HttpKernel/Kernel.php:
```diff
@@ -683,7 +683,7 @@ protected function dumpContainer(ConfigCache $cache, ContainerBuilder $container
 
         if (class_exists('ProxyManager\Configuration')) {
-            $dumper->setProxyDumper(new ProxyDumper());
+            $dumper->setProxyDumper(new ProxyDumper(md5((string) $cache)));
```
Is this unique enough? I know for caching keys Symfony uses sha256 but if this is dumped to a filename it should use sha1.

xphere replied:
This will be called at most twice in a common Symfony application flow, and that's when the cache is empty for the current environment and we call the `cache:clear` command.

The cache path changes between calls, which are (in absolute paths):
- In *dev* environment:
    - app/cache/dev/appDevDebugProjectContainer.php
    - app/cache/de_/ap_DevDebugProjectContainer.php
- In *prod* environment:
    - app/cache/prod/appProdProjectContainer.php
    - app/cache/pro_/ap_ProdProjectContainer.php

I think the risk of collision here is pretty much zero. Choosing `md5` was kind of arbitrary.

I made this PR just to make both executions different. In the future the cache path may not change between calls and we'll need to change the salt. Maybe a simple integer will do: start controlling how many containers the `Kernel` has dumped, and use this index as a salt (even un-hashed). I don't know if it pays off.

fabpot commented:
Thank you @xphere.
